mozilla/fx-private-relay
# MPP-2728: Allow multiple Twilio messaging services

This PR fixes issue #2709:

* Add a test to replicate the issue
* Change `TWILIO_MESSAGING_SERVICE_SID` to a comma-separated value
* Log a 412 response / [21714 errors](https://www.twilio.com/docs/api/errors/21714) when assigning a number to a pool, and try the next SID. This will allow us to pre-allocate the next messaging pool and avoid user signup issues.
* Log 409 response / [21710 errors](https://www.twilio.com/docs/api/errors/21710) when assigning an already assigned number to a pool. This was suggested by @say-yawn in the review for PR #2614.
* Skip registration and log a warning if `TWILIO_MESSAGING_SERVICE_SID` is undefined. This is new behavior - previously, it would be a runtime exception. It minimizes changes to existing tests, which expect to pass with an undefined `TWILIO_MESSAGING_SERVICE_SID` and a mocked twilio client.

Additional changes:

* Address flake8 linting issues
* Rename `mocked_twilio_client` fixture to `mock_twilio_client`, and remove unneeded copying of the mock object. This was also suggested in PR #2614.
* Update `mock_twilio_client` fixture to implement the subset of the client needed for tests
* Add `real_phone_us` fixture to consolidate setup for phone mask creation
* Drop some duplicate test assertions

## How to test

Unit tests should be sufficient. Manual testing can be done by:

* Update `TWILIO_MESSAGING_SERVICE_SID` to two comma-separated values - a full pool SID, and a SID with room
* Create or update a user to have the phone feature and a registered true phone
* On the website, select a relay phone mask:
   - A log message is sent that a message pool is full
   - The relay phone mask was allocated to the second SID
   - The user proceeds to the next step in setting up their phone mask
